### PR TITLE
BUS VOLTAGE: widen window for bus voltage estimate

### PR DESCRIPTION
**Pull request by DSsoto**
* Implement a BusVoltageEstimator class which for now just takes an average
  of the readings from the last n seconds from all thrusters together
* Refactor code related to bus voltage to use the new BusVoltageEstimator
* Moved around functions and renamed variables, functions to imporve readability /
  keep functions with logically close functionality together
* Update sub8_videoray_m5_thruster readme
* Addresses #238

### CODE REVIEW

File: drivers/sub8_videoray_m5_thruster/nodes/thruster_driver.py [hunk 1]
```diff
@@ -14,9 +14,58 @@
...
+    def get_last_update_time(self):
+        ''' Returns time of most recent reading in buffer '''
+        if self.last_update_time is None:
+            self.last_update_time = rospy.Time.now()
+        return self.last_update_time
 
 class ThrusterDriver(object):
-    _dropped_timeout = 1 # s
+    _dropped_timeout = 1.0 # s
+    _window_duration = 30.0 # s
```
Comment (kev-the-dev) [inline drivers/sub8_videoray_m5_thruster/nodes/thruster_driver.py:68]: I feel like 30 seconds is kind of big. We still do want it to kill quickly if the battery drops. Something like 10s seems more reasonable.

File: drivers/sub8_videoray_m5_thruster/readme.md [hunk 2]
```diff
@@ -4,16 +4,17 @@ VideoRay M5 Thruster Driver
 # Authors
 
 This code was originally authored by Forrest Voight (forrestv) and Matthew Griessler.
+It is now maintained by David Soto.
 
 # Purpose
 
-This node handles the interface between the sub's control computer and the thrusters. It communicates via serial to each thruster, individually issuing commands.
+This node handles the interface between the sub's control computer and the thrusters. It communicates via serial to each thruster, individually issuing commands, it also receives response packets from each of th thrusters (including current, voltage, rpm, faults, etc.) and publishes them.
```
Comment (kev-the-dev) [inline drivers/sub8_videoray_m5_thruster/readme.md]: > th thrusters

**Review by kev-the-dev — APPROVED** [review]
The window seems a little big to me. We're really still looking to just throw out an outlier reading right? We still want to kill pretty quickly if the battery is actually low.

File: drivers/sub8_videoray_m5_thruster/nodes/thruster_driver.py (hunk 1 above)
Comment (DSsoto) [reply drivers/sub8_videoray_m5_thruster/nodes/thruster_driver.py:68]: I disagree. I can see it killing after a 10 seconds where we would have come right back above the kill voltage at the end of some move.

However, running it for 20 extra seconds after we actually get to a killable (steady state) voltage does not make much of a difference.

**Comment by DSsoto** [comment]
The bus voltage actually drops momentarily during transient periods with high thrust output, but then recovers after it. It's more like something to filter transient periods than a single reading outlier remover.

**Comment by sentree** [comment]
The way I understand our batteries (which could be completely wrong) is that there is a cutoff voltage where we should stop using them. Even if they only dip below that voltage while moving, isn't that still bad for the batteries in the long term?

**Comment by DSsoto** [comment]
We can ask @duggerd what the optimal value is. I think as it is right now, there is a safety margin above that value where it is harmful for the batteries. I've had the sub kill at about 45.5 volts because it momentarily dipped below 44 volts (kill voltage).

**Comment by DSsoto** [comment]
Merging since there have been no further comments, and we can always just modify the kill voltage.

**Pull request merged**